keps/sig-api-machinery/0014-garbage-collection.md

```diff
@@ -0,0 +1,555 @@
+
+The steps are as easy as:
+
+1. In the ConfigMap and Secret’s metadata, the user sets TTL to 1 hour and that
+   the ConfigMap and Secret should be cleaned up after they are no longer used
+   by other resources (Pods, ReplicaSets, Deployments, etc.)
+1. The user then deploys applications, including ConfigMaps and Secrets of those
+   applications, as usual. Users can create new ConfigMaps and Secrets when they
+   need to change the configuration of their applications. They don’t need to
+   manually check which ConfigMaps and Secrets are not used and can be cleaned
+   up. If the ConfigMaps and Secrets are shared among applications, other
+   applications can still use them before TTL expires. 
+
+#### Story 3: CustomResourceDefinitions (CRDs) and custom resources
+
+Users may create [custom resources][] in their Kubernetes clusters. Users can
+handpick one or more types of garbage collection for each of their custom
+resources and even the definition of those resources
+([`CustomResourceDefinitions`][]/CRDs). For example, a custom resource such as
+Build, Workflow or Function can be cleaned up after they complete or are not in
+use. The TTL provides time for asynchronous clients to observe their final state
+before they are deleted.
+
+[custom resources]: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/
+[`CustomResourceDefinitions`]: https://kubernetes.io/docs/tasks/access-kubernetes-api/extend-api-custom-resource-definitions/
+
+### Implementation Details
+
+Implementation details of the three main parts of the design are discussed
+elsewhere:
+
+1. Object-lifecycle-respecting TTL mechanism: [detailed design of TTL][].
+1. Garbage collect resources that finish execution: [detailed design of GC after
+   use (part 1)][].
+1. Garbage collect resources that are not used by other resources: [detailed
+   design of GC after use (part 2)][].
+
+### Risks and Mitigations
+
+The changes to garbage collector will affect the whole system and need to be
+carefully tested. Tests including, but not limited to:
+
+* Basic tests for each of the garbage collection features:
+  * Resources can be garbage collected.
+  * Resources that opt-out of garbage collection shouldn’t be deleted.
+* When this feature is enabled and disabled, resources won’t be garbage
+  collected.
+* All garbage collection mechanisms can work coherently.
+* Works in clusters with local timezone that changes +/-1hr for DST. 
+
+This feature will also be feature gated.
+
+## Graduation Criteria
+
+This can be promoted to beta when it satisfies users' need for cleaning up
+resources after use, without regressions.
+
+This will be promoted to GA once it's gone a sufficient amount of time as beta
+with no changes. 
+
+## Implementation History
+
+TBD
+
+## Future Work
+
+### Generic Capacity-based Garbage Collection
+
+We may potentially want to support generic capacity-based garbage collection.
```

Comment:
I think the main reason behind supporting capacity-based GC is the users specify too-big TTL values for TTL GC. If so, perhaps we can look at embedding an automated version of capacity-based GC into the TTL garbage collector. In other words, the TTL garbage collector deletes resources when the quota is reached _regardless_ whether TTL value has expired of the resources.

For this to be implemented with correct semantics, the TTL garbage collector should auto-delete the resources with highest TTL value.